### ISSUE TITLE
could not broadcast input array from shape (105) into shape (50)

### ISSUE THREAD

**Issue by ElBart00** [EpistasisLab/tpot#1142]
I have a panda's dataset where I want to find the best classifier machinelearning algorithm for.

RangeIndex: 7820 entries, 0 to 7819
Data columns (total 6 columns):
    Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   CategoryName          7820 non-null   object
 1   CategoryId            7820 non-null   int64 
 2   ManufacturerCode      7820 non-null   object
 3   ManufacturerCategory  7820 non-null   object
 4   Description           7820 non-null   object
 5   Description2          7820 non-null   object

I've created a preprocessor to transform the text data to numeric data:
`hash_transformer = Pipeline(steps=[
    ('hasher', FeatureHasher(n_features=2**17, input_type='string')),
])

vector_transformer = Pipeline(steps=[
    ('vector', HashingVectorizer(n_features=2**20)),
])

preprocessor = ColumnTransformer(
    transformers=[
        ('manufacturerCode', hash_transformer, 'ManufacturerCode'),
        ('description', vector_transformer, 'Description'), 
        ('description2', vector_transformer, 'Description2'),
        ('ManufacturerCategory', vector_transformer, 'ManufacturerCategory'), 
    ], n_jobs=-1, verbose=True)
`

When I try to run TPOT I get an error:
`X = preprocessor.fit_transform(data)
y = data['CategoryId'].values

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
tpot = TPOTClassifier(config_dict='TPOT sparse', n_jobs= -1, population_size=2, generations=2, verbosity=0, random_state=42)
tpot.fit(X_train, y_train)
`

This is the whole error message:

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-6-da7b509edb1d> in <module>
      5 tpot = TPOTClassifier(config_dict='TPOT sparse', n_jobs= -1, population_size=2, generations=2, verbosity=0, random_state=42)
      6 
----> 7 tpot.fit(X_train, y_train)
      8 print(tpot.score(X_test, y_test))
      9 #tpot.export('tpot_digits_pipeline.py')

c:\users\b.vanlier\source\bomlist jupyter\.env\lib\site-packages\tpot\base.py in fit(self, features, target, sample_weight, groups)
    659         features, target = self._check_dataset(features, target, sample_weight)
    660 
--> 661         self._init_pretest(features, target)
    662 
    663         # Randomly collect a subsample of training samples for pipeline optimization process.

c:\users\b.vanlier\source\bomlist jupyter\.env\lib\site-packages\tpot\tpot.py in _init_pretest(self, features, target)
     62         if not np.array_equal(np.unique(target),np.unique(self.pretest_y)):
     63             unique_target_idx = np.unique(target,return_index=True)[1]
---> 64             self.pretest_y[0:unique_target_idx.shape[0]] = \
     65                     _safe_indexing(target, unique_target_idx)
     66 

ValueError: could not broadcast input array from shape (105) into shape (50)

It seems it has to do somenthing with the input data, but I can't get a grasp on what that might be. When I limit the input data to the first 4000 rows it works fine. When I try the first 5000 rows I get the error. When I use the rows between 4000 and 5000 it works fine. My first tought was some bogus characters in the input data but I cant find any consistencies to track these records down.
Any help on resolving this would be apreciated.

**Comment by weixuanfu** [comment]
How many unique classes in y?

-------------------------------------------------------------------------------
Weixuan Fu

Email: weixuanf@pennmedicine.upenn.edu<mailto:weixuanf@pennmedicine.upenn.edu>


On Dec 15, 2020 at 6:24 AM, <Bart<mailto:notifications@github.com>> wrote:


I have a panda's dataset where I want to find the best classifier machinelearning algorithm for.

RangeIndex: 7820 entries, 0 to 7819
Data columns (total 6 columns):

Column Non-Null Count Dtype
________________________________

0 CategoryName 7820 non-null object
1 CategoryId 7820 non-null int64
2 ManufacturerCode 7820 non-null object
3 ManufacturerCategory 7820 non-null object
4 Description 7820 non-null object
5 Description2 7820 non-null object

I've created a preprocessor to transform the text data to numeric data:
`hash_transformer = Pipeline(steps=[
('hasher', FeatureHasher(n_features=2**17, input_type='string')),
])

vector_transformer = Pipeline(steps=[
('vector', HashingVectorizer(n_features=2**20)),
])

preprocessor = ColumnTransformer(
transformers=[
('manufacturerCode', hash_transformer, 'ManufacturerCode'),
('description', vector_transformer, 'Description'),
('description2', vector_transformer, 'Description2'),
('ManufacturerCategory', vector_transformer, 'ManufacturerCategory'),
], n_jobs=-1, verbose=True)
`

When I try to run TPOT I get an error:
`X = preprocessor.fit_transform(data)
y = data['CategoryId'].values

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
tpot = TPOTClassifier(config_dict='TPOT sparse', n_jobs= -1, population_size=2, generations=2, verbosity=0, random_state=42)
tpot.fit(X_train, y_train)
`

This is the whole error message:

________________________________

ValueError Traceback (most recent call last)
in
5 tpot = TPOTClassifier(config_dict='TPOT sparse', n_jobs= -1, population_size=2, generations=2, verbosity=0, random_state=42)
6
----> 7 tpot.fit(X_train, y_train)
8 print(tpot.score(X_test, y_test))
9 #tpot.export('tpot_digits_pipeline.py')

c:\users\b.vanlier\source\bomlist jupyter.env\lib\site-packages\tpot\base.py in fit(self, features, target, sample_weight, groups)
659 features, target = self._check_dataset(features, target, sample_weight)
660
--> 661 self._init_pretest(features, target)
662
663 # Randomly collect a subsample of training samples for pipeline optimization process.

c:\users\b.vanlier\source\bomlist jupyter.env\lib\site-packages\tpot\tpot.py in _init_pretest(self, features, target)
62 if not np.array_equal(np.unique(target),np.unique(self.pretest_y)):
63 unique_target_idx = np.unique(target,return_index=True)[1]
---> 64 self.pretest_y[0:unique_target_idx.shape[0]] =
65 _safe_indexing(target, unique_target_idx)
66

ValueError: could not broadcast input array from shape (105) into shape (50)

It seems it has to do somenthing with the input data, but I can't get a grasp on what that might be. When I limit the input data to the first 4000 rows it works fine. When I try the first 5000 rows I get the error. When I use the rows between 4000 and 5000 it works fine. My first tought was some bogus characters in the input data but I cant find any consistencies to track these records down.
Any help on resolving this would be apreciated.

**Comment by ElBart00** [comment]
107

`print(pd.Series(y).unique())
print(len(pd.Series(y).unique()))`

[ 18  14   5  16   6  71   8  30  29  25  20  27  35  28  59  91  36  33
  15  95  23  22   9  34  32  26  11  13  19  12  31  24   7  17   4 100
  39 101  81  57  55  56  21  42  46  72  68 118  74  75  44 114  80  93
  98  41  89  49  78  86  88  87  90 106  52  79  51  58  60  43  92 104
  48 103  64  97  45 105  38  40 109 107 113 110  47  85  76  84  83  77
 102  73  70  63  54  53  50  94 108 111  99 116 115 119  69 112 117]
107

**Comment by weixuanfu** [comment]
In this case, I think it will trigger an issue when you are using TPOTClassifier if it is indeed not a regression problems. 

TPOT should try to randomly sample a subset of 50 samples that can covers all unique y values for pre-testing pipelines. But since it is more than 50 unique values, then that down-sampling step is not working correctly. 

We will fix this issue.

**Comment by weixuanfu** [comment]
I fixed it via PR #1143  and it was merged to development branch. And It will be included in next release of TPOT.

For testing the development branch, you may install TPOT with patch into your environment via:
```
pip install --upgrade --no-deps --force-reinstall git+https://github.com/EpistasisLab/tpot.git@development
```

**Comment by ElBart00** [comment]
Thanks for the support!
The fix seems to work.
